src/sgbuf.c

```diff
@@ -99,23 +99,56 @@ size_t trace_indices( sgIdx_t *idx, size_t len, struct trace tr) {
         cur += i;
     }
     assert (cur == len);
-
+    // Pass over sidx[], convert byte addresses to indicies, track min.
     sidx[0] /= 8;
     sgsIdx_t min = sidx[0];
     for (size_t i = 1; i < len; i++) {
         sidx[i] = sidx[i-1] + sidx[i] / 8;
         if (idx[i] < min) 
             min = sidx[i];
     }
-
+    // Translate to zero-based start index, track max.
     idx[0] = sidx[0] - min;
     size_t max = idx[0];
     for (size_t i = 1; i < len; i++) {
         idx[i] = sidx[i] - min;
         if (idx[i] > max) 
             max = idx[i];
     }
-
+    // Pageinate the positive zero-based indicies in idx[].
+    long *pages = NULL, npages = 0;
+    long  page,pidx;
+    long  page_bits = 26; // 26 => 64MiB
+    long  new_idx;
+    long  new_max = 0;
+    for(size_t i = 0; i < len; i++) {
+      // Turn address into page.
+      page = (idx[i]*8) >> page_bits;
+      // Find existing / make new page entry.
+      pidx = -1;
+      for(size_t p = 0; p < npages; p++) {
+	if( pages[p] == page ) {
+	  pidx = p;
+	}
+      }
+      if( pidx == -1 ) {
+	pidx = npages;
+	npages++;
+	if( !(pages = realloc(pages,npages*sizeof(long))) ) {
+	  fprintf(stderr,"trace_indices(): Failed to allocate new page entry (%ld).\n",npages);
+	}
+	pages[pidx] = page;
+      }
+      // Replace sparse page bits in address with dense page index bits.
+      new_idx  = (pidx << page_bits) | ((idx[i]*8) & ((1l<<page_bits)-1l));
+      new_idx /= 8;
+      idx[i] = new_idx;
+      if( idx[i] > new_max ) {
```

Comment:
I believe your code should have the ability to reduce the amount of memory required, correct? In that case, should we check to see if `max` can be lowered at the end of your code?

Reply:
Besides that, the code looks good.

Reply:
Yes, the idea to have max lowered. Here is an example:

./spatter -l 1000000 -k gather -b OpenMP -config-file ../../app_traces_stats_v2/amg.pat 
page compression: 18446744073709551582 25281151

That is, without compacting touched pages, one would need 18446744073709551582*8 bytes, but when only using touched pages and compacting them you only need 25281151*8 bytes. IDK, might be off slightly here, but I think you get the idea. Without this, the memory requirements are intractable.

Reply:
That's *8 at the end of both of those btw.

Reply:
Sorry, I misread a bit of the code. Looks good.